### PR TITLE
Contributing towards issue #419 "Requirements for list of mentors/students"

### LINKED ISSUE #419: Requirements for lists of mentors/students
### Mentor list (for org admins/sub-org admins):
- name
- email 
- sub-org(s)
- alternate contact methods (irc, phone, etc.) (maybe click to profile for this because of width limits)
- do they have an invite to google's system or not?

This list should be current year only by default, with maybe an option to see previous year's list if you're an org admin.

**Things org admins might want to modify with mentors**: 
- add/subtract sub-orgs (maybe not necessary if we fix the bug around this)
- tick the tickybox when they get an invite (solved)

**Things org admins want**: 
- get list of emails for all mentors per sub org (for emails not sent through the system) 
- get list of emails for all mentors (for syncing with mailman) -- currently getting that from teh google csv export.

### Student list (for mentors/sub-org admins)
- name
- email (maybe not needed)
- sub-org
- blogging compliance status (e.g. are they late on any blog posts?  have they already posted the one for this week?)
- link directly to blog

Again, current year only by default, option to see previous year's lists.

Right now this completely doesn't exist, so we'd need to make sure all mentors have access to this.

### PR DESCRIPTION
**Pull request by chandan00761**
# Commits

+added GSoCYearFilter which filters users by year with current year as default value.

! changed get_queryset of HiddenUserProfileAdmin such that:
  * when request.user is superadmin all users are returned
  * when request.user is Suborg Admin users of that suborg are returned.
  * else empty query is returned.

# Description

[changes described here](https://github.com/python-gsoc/python-blogs/issues/419#issuecomment-982673691)

Fixes #419 

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update.

# How Has This Been Tested?

**To make this feature work do the following steps:**
* Create a new group suborg_admin and provide permission "gsoc | user profile | Can view user profile".
* Add all the suborg admins to this group.

**Testing**
* Log in as any of the suborg admin and go to the admin screen.
* Pick **User Profiles** under **GSoC**.

- [x] Manual Testing

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have not added a commit to any .db files as part of my pull request
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: gsoc/admin.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 from .models import *
+from .filters import GSoCYearFilter
```
Comment (botanicvelious) [inline gsoc/admin.py]: dont need to make a whole new file and class for this

File: gsoc/admin.py [hunk 2]
```diff
@@ -396,13 +397,23 @@ def email(self, obj):
...
-        return UserProfile.all_objects.all()
+        """
+        This queryset is used to return UserProfile object based on the role of request.user
+        If the user is a super user then all users are returned.
+        If the user is a suborg admin then mentors and students of that suborg are returned.
+        """
+        user = request.user.userprofile_set.get()
+        if request.user.is_superuser:
+            return UserProfile.all_objects.all()
+        if user.role == 1:
```
Comment (botanicvelious) [inline gsoc/admin.py]: so a user can have more then one role, so you will need to check all roles for the current year

File: gsoc/filters.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
+from datetime import datetime
+
+from django.utils.translation import gettext_lazy as _
+from django.contrib.admin import SimpleListFilter
+
+
+class GSoCYearFilter(SimpleListFilter):
```
Comment (botanicvelious) [inline gsoc/filters.py]: This entire class isnt really necessary, no need to make a new file and class for one function, just stick this code in admin.py also doing it this way is like 2x the amount of code

File: gsoc/admin.py [hunk 4]
```diff
@@ -396,13 +397,23 @@ def email(self, obj):
...
+            return UserProfile.objects.filter(suborg_full_name=user.suborg_full_name)
```
Comment (botanicvelious) [inline gsoc/admin.py]: you could set this filter to be the current year, as well as the suborg in question at the same time, doing so would make the filters.py redundant

File: gsoc/admin.py [hunk 5]
```diff
@@ -353,10 +354,10 @@ class HiddenUserProfileAdmin(admin.ModelAdmin):
         "hidden",
         "reminder_disabled",
         "current_blog_count",
-        "role", 
+        "role",
         "gsoc_invited",
     )
-    list_filter = ("hidden", "reminder_disabled", "role", "gsoc_invited")
+    list_filter = ("hidden", "reminder_disabled", "role", "gsoc_invited", GSoCYearFilter)
```
Comment (botanicvelious) [inline gsoc/admin.py]: As I mentioned, why have a new class? just display the data that is already there.

File: gsoc/admin.py [hunk 6]
```diff
@@ -396,13 +397,23 @@ def email(self, obj):
...
     def get_queryset(self, request):
-        return UserProfile.all_objects.all()
+        """
+        This queryset is used to return UserProfile object based on the role of request.user
+        If the user is a super user then all users are returned.
+        If the user is a suborg admin then mentors and students of that suborg are returned.
+        """
+        user = request.user.userprofile_set.get()
+        if request.user.is_superuser:
+            return UserProfile.all_objects.all()
```
Comment (botanicvelious) [inline gsoc/admin.py]: this should be limited to the current year

**Review by botanicvelious — CHANGES_REQUESTED** [review]
Let me know if you have any questions on my comments, for something like this its usually best to not make extra classes etc that will only be used once, especially when it can be done in the same filter query you are already using

**Comment by chandan00761** [comment]
Thank you for the feedback. I implement the feature in a more "beginner friendly" way hence more lines of code. I will make all the changes suggested and ping you back.

Changelist:

- [x] remove filters.py and GSoCYearFilter
- [x] add inline year filter.
- [x] check roles of all UserProfiles
- [x] All users of only current year should be shown to admin by default.

File: gsoc/admin.py (hunk 1 above)
Comment (chandan00761) [reply gsoc/admin.py]: Fixed

File: gsoc/filters.py (hunk 3 above)
Comment (chandan00761) [reply gsoc/filters.py]: Fixed

File: gsoc/admin.py (hunk 2 above)
Comment (chandan00761) [reply gsoc/admin.py]: if the user is superuser then we return all the UserProfiles sorted by latest year.
```
UserProfile.all_objects.all().order_by("gsoc_year")
```
else we check if any of the UserProfiles has role set as suborg admin
```
user_profiles = user.userprofile_set.filter(role=1).all()
```

File: gsoc/admin.py (hunk 4 above)
Comment (chandan00761) [reply gsoc/admin.py]: For suborg admins only current year mentors are shown for psf admins all the users are shown sorted by year.

File: gsoc/admin.py (hunk 5 above)
Comment (chandan00761) [reply gsoc/admin.py]: removed filter

**Comment by chandan00761** [comment]
I have done the changes requested. Please review if it gets the work done.

**Comment by botanicvelious** [comment]
Took a quick look, the overall logic looks right now, just 2 questions

1) why is this a change to get_queryset in the HiddenUserProfileAdmin class?
2) how will users be able to see this list? should probably add a menu item for suborg admins for it

**Comment by chandan00761** [comment]
1. Since only a few conditions needed checking to return the filtered Queryset, I have done it inside the get_queryset only.
2. Suborg admin can access the list from UserProfiles menu item only. Only thing they should have is, access to the Admin dashboard and  "gsoc | userprofiles | can view user profile" permission.

**Comment by botanicvelious** [comment]
My concern with editing the other class is that class may be used elsewhere, I dont see it being used anywhere however so not sure why it existed before but im likely missing it.

For the permission that is required then we need code added for adding the permission automatically dont we? were you able to test the code?

**Comment by chandan00761** [comment]
@botanicvelious

**Pull request merged**